Treat address labels as "const" symbols

antis81 opened the pull request:
This PR serves as a basis to check on symbol parsing states during passes. States are:

- ~~declared  -> Symbol has been registered (with or without a value assigned)~~
- accessed -> Symbol is accessed from at least one code location (not necessarily defined)
- defined    -> Symbol is declared and value has been assigned

Other attributes:
- final         -> Symbol is marked constant (redefining it leads to error)

antis81 committed:
throw if a const label has been defined already

Note: Currently outputs the complete label string.

antis81 committed:
use ::Check() function to check certain label pre-conditions

antis81 committed:
simplify string assignment from string_view

sasq64 commented on src/assembler.cpp:
```diff
@@ -443,22 +443,21 @@ void Assembler::handleLabel(std::any const& lbl)
             lastLabel = std::any_cast<std::string_view>(lbl);
         }
     }
+    ::Check(syms.is_defined(label) == false, fmt::format("Symbol '{}' already declared", label));
```
Sadly this doesn't work -- any time more than one pass is required this will cause an error.

antis81 replied:
Alright, wasn't aware this is "layered".

antis81 replied:
Removed the commit then.

sasq64 reviewed (commented):
It is a problem that duplicate labels does not give an error, its on my TODO... but it's not this easy to fix :)
(Labels are redefined in subsequent passes)

antis81 committed:
rename is_defined() -> is_declared()

antis81 committed:
reintroduce is_defined()

Now checks on the Symbols "defined" flag.
Note: A defined symbol is implicitly declared.

antis81 commented:
@sasq64 Giving it another try… :) Could you please give a feedback on the latest 2 commits moving this in the right direction? Why would subsequent passes try to redefine a symbol once it is complete?

antis81 committed:
add SymbolTable::is_accessed()

antis81 commented:
Ran the existing tests manually via `make test` -> :ok:

Are you ok with changing the meaning of "is_defined" to "is_declared"?

antis81 commented:
@sasq64 Could you please double-check commit b6a5674?

sasq64 commented:
Sorry for not answering eariler,
Anyway my only problem is that is not clear what "declared" means, since you can't decalare and not define symbols.
Isn't it really the same as "accessed" ?
But if we need to keep it, I think it should be renamed to `exists()` or `used()`.

antis81 committed:
remove is_declared (slightly changes the is_defined logic)

antis81 committed:
call set_sym<AnyMap>() directly on a map

Before this call was redirected: set<AnyMap>() -> set_sym<AnyMap>().
Suggestion: rename "set_sym" to "set_map" making it more clear.

antis81 commented:
First the easy stuff. Yeah you're right the endless recursion does not happen -> cleaned up a little by calling "set_sym(name,map)" directly (1ed1b68). Rename to "set_map" may help.

Went through the get/set functions in detail yesterday -> Making it short: Yeah, "accessed" and "defined" should be enough state(s). In current implementation the "is_defined" checks on the "defined" flag in addition to "exists()" -> test 65c02 fails (A != 1 -> which should be in table[0])

Looks like one of the "is_defined" needs to change to "is_accessed"… needs a bit more investigation/debugging.

I like to go back to the original intention creating this issue -> check if a (constant) label had been redefined. This should be the "final" attribute right? This should apply (only?) on "pure" address labels (label without an "=" assignment operator). Ah… code is easier to understand:

```asm
!org $1000  ; just to give a section context
i_am_var = 42  ; a "var" label type 
i_am_label:      ; a "final" label type
```

Shouldn't be overly complex - am I missing something?

antis81 committed:
add a question on smybol(s) defined/undefined

antis81 committed:
A symbol is defined when in "syms" and *not* in "undefined".

note: The "defined" flag is ignored (probably no longer required)!

antis81 commented:
@sasq64 FYI: The tests are ok again ("table" is defined). Seems the "defined" flag is not always set/cleared correctly - but the lists are. Can we remove the "defined" flag completely?

antis81 committed:
remove defined flag

antis81 committed:
set const labels "final" -> error out if redefined

antis81 commented:
Hi @sasq64,
reached the "works for me" state! The test `./tests/err_labels.asm` now fails (can it give an "ok" result instead aka "death test" or something?

```bash
builds/debug/tester all
/home/nils/Projects/badass/./tests/65c02.asm
* PARSING
Using cached AST
* PASS 1
* PASS 2
* TESTS (1)
Running code at $1000
Start
*** Test 'start' : 93 cycles, [A=$02 X=$02 Y=$03]
/home/nils/Projects/badass/./tests/banks.asm
* PARSING
Using cached AST
* PASS 1
* TESTS (1)
Running code at $2000
BANK    2
BANK    1
BANK    2
BANK    1
*** Test 'test' : 307 cycles, [A=$01 X=$ff Y=$10]
/home/nils/Projects/badass/./tests/basic_errors.asm
* PARSING
Using cached AST
* PASS 1
ERROR 'Illegal instruction 'xyz'' in /home/nils/Projects/badass/./tests/basic_errors.asm:4
/home/nils/Projects/badass/./tests/copy_test.asm
* PARSING
Using cached AST
* PASS 1
* PASS 2
* TESTS (2)
Running code at $800
*** Test 'my_test' : 193 cycles, [A=$a9 X=$0c Y=$00]
Running code at $1007
*** Test 'dummy' : 5 cycles, [A=$04 X=$00 Y=$00]
* FINAL PASS
$a9
…
... <all tests before tests/err_labels.asm are OK> …
…
/home/nils/Projects/badass/./tests/err_labels.asm                                                                                                                                                                                                                             
* PARSING                                                                                                                                                                                                                                                                     
Using cached AST
* PASS 1
ERROR 'Const Label 'same_label' cannot be redefined' in /home/nils/Projects/badass/./tests/err_labels.asm:4
                                                                                                                                                                                                                                                                              
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
tester is a Catch v2.7.0 host application.
Run with -? for options                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                              
-------------------------------------------------------------------------------                                                                                                                                                                                               
all
-------------------------------------------------------------------------------
/home/nils/Projects/badass/src/asmtest.cpp:164
...............................................................................                                                                                                                                                                                               

/home/nils/Projects/badass/src/asmtest.cpp:175: FAILED:
explicitly with message:                                                                                                                                                                                                                                                      
  Did not find expected errors

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed

```

antis81 committed:
fixup wording for test

antis81 committed:
fix include test

sasq64 commented:
Testing this now.
I am afraid an important case may not be covered by the tests, because labels  _can_ move, exampe:
lda something,x
label:
    nop
    
If something < 0 its a different opcode than > ff so label will be moved 1 byte for next pass
